buckets-mdapi/src/error.rs

```diff
@@ -1,58 +1,84 @@
     fn to_string(&self) -> String {
         match *self {
-            BucketsMdapiErrorType::BucketAlreadyExists => "BucketAlreadyExists".into(),
-            BucketsMdapiErrorType::BucketNotFound => "BucketNotFound".into(),
-            BucketsMdapiErrorType::ObjectNotFound => "ObjectNotFound".into(),
-            BucketsMdapiErrorType::PostgresError => "PostgresError".into(),
-            BucketsMdapiErrorType::LimitConstraintError => "LimitConstraintError".into(),
+            BucketsMdapiError::BucketAlreadyExists => "BucketAlreadyExists".into(),
+            BucketsMdapiError::BucketNotFound => "BucketNotFound".into(),
+            BucketsMdapiError::ObjectNotFound => "ObjectNotFound".into(),
+            BucketsMdapiError::LimitConstraintError(_) => "LimitConstraintError".into(),
+            BucketsMdapiError::PreconditionFailedError(_) => "PreconditionFailedError".into(),
+            BucketsMdapiError::PostgresError(_) => "PostgresError".into(),
+            BucketsMdapiError::ContentMd5Error(_) => "ContentMd5Error".into(),
         }
     }
 }
 
-impl BucketsMdapiErrorType {
-    fn message(self) -> String {
+impl BucketsMdapiError {
+    pub fn message(&self) -> String {
         match self {
-            BucketsMdapiErrorType::BucketAlreadyExists => "requested bucket already exists".into(),
-            BucketsMdapiErrorType::BucketNotFound => "requested bucket not found".into(),
-            BucketsMdapiErrorType::ObjectNotFound => "requested object not found".into(),
-            BucketsMdapiErrorType::PostgresError => "postgres encountered an error".into(),
-            BucketsMdapiErrorType::LimitConstraintError => "a limit constraint was violated".into(),
+            BucketsMdapiError::BucketAlreadyExists => "requested bucket already exists".into(),
+            BucketsMdapiError::BucketNotFound => "requested bucket not found".into(),
+            BucketsMdapiError::ObjectNotFound => "requested object not found".into(),
+            BucketsMdapiError::LimitConstraintError(msg) => msg.to_string(),
+            BucketsMdapiError::PreconditionFailedError(msg) => msg.to_string(),
+            BucketsMdapiError::PostgresError(msg) => msg.to_string(),
+            BucketsMdapiError::ContentMd5Error(msg) => {
+                format!("content_md5 is not valid base64 encoded data: {}", msg)
+            },
         }
     }
+
+    /*
+     * The Fast protocol expects our errors to look like so:
+     *
+     *     { "error": { "name": "...", "message": "..." } }
+     *
+     * We can get most of the way there by annotating BucketsMdapiError with Serde's tag/content
+     * for enum {de,}serialization, but it falls a bit short in its current form because not all
+     * elements include a String variant.  This means that "message" is either not present while
+     * serializing, or causes an error while deserializing.
+     *
+     * Paired with the fact that we also need the outer "error" property, it seems the easiest way
```

Comment:
Yeah, a lot of this problem is my fault for the way I specified the `rust-fast` interface for creating fast message. Rather than requiring `Value` types they should instead be generic over types that implement `Deserialize` and that would avoid a lot of the hassle here with trying to make everything a `Value` type. I'll probably try to make a change to those rust-fast APIs at some point and then we can revisit this and make it a bit nicer.

Reply:
It's no bother. I just thought I'd write this comment up to help me clarify the wrapping a little bit. I wanted to type this comment up to help me understand the wrapping a little bit, but if you think it's out of place I don't mind removing it.

Reply:
No, I think it's perfectly fine to remain until I can make the changes to `rust-fast` and pull that in.